File: javascript/src/linkedlist.js
'use strict';
/**
 * Linked list implementation in vanilla JavaScript.
 * @author Brandon Manke
 * @file linkedlist.js
 */
function LinkedList() {
    this.head = null;
    this.size = 0;

    /**
     * This is a constructor for list elements / nodes.
     * @param data - data that node represents
     */
    function Node(data) {
        this.data = data;
        this.next = null;
    }
};

/**
 * @return size - Returns size of list
 */
LinkedList.prototype.getSize = function() {
    return this.size;
}

/**
 * Returns the value of the data at specified node.
 * @param index - Index of node, whose value is returned.
 * @return value - Value of node at specified index. (Or -1 if not found)
 */
LinkedList.prototype.getValue = function(index) {
    var current = this.head;
    var count = 0;
    if (index < 0 || index > (this.size - 1)) {
        throw console.log('Index out of bounds');
    }

    while (current.next !== null) {
        if (count === index) {
            return current.data;
        }
                
        if (count < this.size - 1) {
            count++;
        }
        current = current.next;
    }
    return -1; // if nothing is found
}

/**
 * Adds node at the beginning of the list.
 * @param data - data to be added to list
 */
LinkedList.prototype.insertBefore = function(data) {
    var current = new Node(data);

    if (this.head === null) {
        this.head = current;
        this.size++;
        return;
    }

    current.next = this.head;
    this.head = current;
    this.size++;
}

/**
 * Adds node at the end of the list.
 * @param data - data to be added to list
 */
LinkedList.prototype.insertAfter = function(data) {
    var current = this.head;

    if (this.head === null) {
        this.head = new Node(data);
        this.size++;
        return;
    }

    while (current.next !== null) {
        current = current.next;
    }

    current.next = new Node(data);
    this.size++;
}

/**
 * Removes node at the beginning of the list.
 * @return data of removed node
 */
LinkedList.prototype.removedHead = function() {
    if (this.head === null) {
        throw console.log('No such element');
    }
    var temp = this.head;
    this.head = temp.next;
    this.size--;
    return temp.data;
}

/**
 * Removes node at end of the list.
 * @return data of removed node
 */
LinkedList.prototype.removeEnd = function() {
    var current = this.head;
    if (this.head === null) {
        throw console.log('No such element');
    }

    while (current.next.next !== null) {
        current = current.next;
    }

    var temp = current.next;
    current.next = null;
    this.size--;
    return temp.data;
}

/**
 * Removes node at specified index.
 * @param index - position of element to be removed in list
 * @return data of removed node
 */
LinkedList.prototype.remove = function(index) {
    var current = this.head;
    var temp;
    var count = 0;
    if (index < 0 || index > (this.size - 1)) {
        throw console.log('Index out of bounds');
    }

    // TODO finish this logic prev.next = current.next
    while (current.next !== null) {
        if (count === index - 1) {
            temp = current.next;
            current.next = temp.next;
        }

        count++;
        current = current.next;
    }
    return temp.data;
}

/**
 * Reverses linked-list.
 */
LinkedList.prototype.reverse = function() {

}

/**
 * Removes first instance of value in linked-list.
 * @param value - value to be removed
 */
LinkedList.prototype.removeValue = function(value) {

}

/**
 * Returns list formatted as string with '-->' seperating data.
 * @return string - Formatted string of entire list data
 */
LinkedList.prototype.toString = function() {
    var current = this.head;
    var string = '';
    if (this.head !== null) {
        while (current.next !== null) {
            //if (current.next.next === null) {
            //    string += current.data + '-->' + current.next.data;
            //}
            string += current.data + '-->';
            current = current.next;
        }
    } else {
        throw console.log('List is empty');
    }
    string += current.data;
    return string;
}

// Tests:
var LinkedList = new LinkedList();
LinkedList.insertAfter(2);
LinkedList.insertBefore(3);
LinkedList.insertAfter(7);
LinkedList.insertAfter(5);
console.log('Size:', LinkedList.getSize());
console.log(LinkedList.toString()); // Prints 3-->2-->7-->5

File: c++/singly_linked_list/src/linked_list.cpp
/**
 * Simple singly linked list implementation
 * @author Brandon Manke
 */
#include <string>

template <class E>
class LinkedList {
    // list node
    struct Node {
        E data;
        Node *next;
    };

    private:
        Node *head;
        int size;
    public:
        LinkedList() {
            head = 0;
            size = 0;
        }

        ~LinkedList() {
            Node *current = head;
            while (current != 0) {
                Node *next = current->next;
                delete current;
                current = next;
            }
            head = 0;
        }

        /**
         * Adds value to end of linked list.
         * @param data - value to be added to end of list
         */
        void addToEnd(const E data) {
            if (head == 0) {
                head = new Node();
                head->data = data;
                head->next = 0;
                size++;
                return;
            }
            Node *current = head;
            while (current->next != 0) {
                current = current->next;
            }
            Node *temp = new Node();
            temp->data = data;
            temp->next = 0;
            current->next = temp;
            size++;
        }

        /**
         * Adds value to head/front of linked list.
         * @param data - value to be added to front of list
         */
        void addToFront(const E data) {
            Node *temp = new Node();
            temp->data = data;
            temp->next = head;
            head = temp;
            size++;
        }

        void addTo(const int index, const E data) {
            int counter = 0;

            if (index == 0)
                addToFront(data);

            if (index == getSize() - 1)
                addToEnd(data);

            Node *temp = new Node();
            temp->data = data;
            Node *current = head;

            while (current->next != 0) {
                if (counter == index - 1) {
                    break;
                }
            }
            temp->next = current->next;
            current->next = temp;
            size++;
        }

        E removeEnd() {
            if (head == 0)
                throw;

            Node *current = head;
            while (current->next->next != 0) {
                current = current->next;
            }
            Node *removed = current->next;
            delete current->next;
            current->next = 0;
            size--;
        }

        E removeFront() {
            if (head == 0)
                throw;

            Node *temp = head;
            head = head->next;
            size--;
        }

        E remove(const int index) {
            if (index == 0)
                removeFront();

            if (index == getSize() - 1)
                removeEnd();

            Node *current = head;
            Node *removed;
            int counter = 0;
            while (current->next != 0) {
                if (counter == index - 1) {
                    removed = current->next;
                    if (removed->next != 0) {
                        current->next = removed->next;
                    } else {
                        current->next = 0;
                    }
                } else {
                    counter++;
                    current = current->next;
                }
            }
        }

        /**
         * Prints list to string
         * @return str - list elements seperated by '->' in a string
         */
        std::string toString() const {
            std::string str = "";
            // returns true if template variable is a char or string at runtime
            bool isCharOrString = (std::is_same<E, char>::value || std::is_same<E, std::string>::value);
            if (head == 0) {
                return "List is empty.";
            }
            Node *current = head;
            while (current->next != 0) {
                // check if char or string, so it isnt converted to ascii value
                if (isCharOrString) {
                    str += current->data;
                    str += "->";
                } else {
                    str += std::to_string(current->data) + "->";
                }
                current = current->next;
            }
            if (isCharOrString) {
                str += current->data;
            } else {
                str += std::to_string(current->data);
            }
            return str;
        }

        /**
         * @return size of list
         */
        int getSize() const {
            return size;
        }

        /**
         * Sums all nodes of the list
         * @return sum - sum of entire list
         */
        int sum() const {
            //need checks for type of (double, float, int, long, etc.)
            int sum = 0;
            if (head == 0) {
                throw "List is empty, cannot sum list.";
            }
            Node *current = head;
            while (current != 0) {
                sum += current->data;
                current = current->next;
            }
            return sum;
        }

        /**
         * Computes average of all the values in list
         * @return average - floating point value of average
         */
        float average() const {
            //need checks for type of (double, float, int, long, etc.)
            float average = 0;
            if (head == 0) {
                throw "List is empty, cannot get average.";
            }
            average = this->sum();
            average /= size;
            return average;
        }

        /**
         * Adds param list to end of this linkedlist
         * @param list - LinkedList reference to be added to end of current list
         */
        void concat(const LinkedList<E> &list) {
            if (head == list.head) {
                return;
            }
            if (head == 0 && list.head != 0) {
                head = list.head;
                size += list.getSize();
                return;
            }
            Node *current = head;
            while (current->next != 0) {
                current = current->next;
            }
            current->next = list.head;
            size += list.getSize();
        }
};